allow show tools to be moved to the header, closes #2316

dnoneill opened the pull request:
Add a new function called add_show_header_tools_partial. This will allow the user to add tools to the page header. 

When the app is spun up nothing changes. Changes only go into effect if the user enables these settings.


<img width="1115" alt="Screenshot 2024-08-15 at 11 10 15 AM" src="https://github.com/user-attachments/assets/c2c9e223-d2c4-4579-92ff-16151459fab4">


add_show_header_tools_partial small screen
<img width="647" alt="Screenshot 2024-08-15 at 11 06 45 AM" src="https://github.com/user-attachments/assets/4d9bf3c8-22c0-4fd3-a0c3-d3b4098b8124">

jcoyne commented:
@dnoneill can you post some screenshots that also have the "more like this" component?

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,49 @@
+        search_context.present? || search_session.present? || has_header_actions
+      end
+
+      def before_render
```
why is this `before_render` Can we make a method `has_header_actions?` (and remove the `attr_reader`)

dnoneill replied:
done.

jcoyne commented on app/helpers/blacklight/component_helper_behavior.rb:
```diff
@@ -37,6 +37,10 @@ def document_actions(document, options: {})
       filter_partials(blacklight_config.view_config(:show).document_actions, { document: document }.merge(options)).map { |_k, v| v }
     end
 
+    def header_actions(document, options: {})
```
Is there a reason we need this helper?  Can we just have it in the component?  We've been trying to move away from helper methods.

dnoneill replied:
Done.

jcoyne commented on app/helpers/blacklight/component_helper_behavior.rb:
```diff
@@ -37,6 +37,10 @@ def document_actions(document, options: {})
     end
 
+    def header_actions(document, options: {})
+      filter_partials(blacklight_config.view_config(:show).header_actions, { document: document }.merge(options)).map { |_k, v| v }
```
Is it possible to do?
```suggestion
      filter_partials(blacklight_config.view_config(:show).header_actions, { document: document }.merge(options)).values
```

dnoneill replied:
Nope causes an error. Something about how the ActionsComponent is set up doesn't like it.

jcoyne commented on app/views/catalog/_show_main_content.html.erb:
```diff
@@ -1,4 +1,4 @@
-<%= render blacklight_config.track_search_session.item_pagination_component.new(search_context: @search_context, search_session: search_session, current_document: @document) if blacklight_config.track_search_session.item_pagination_component %>
+<%= render Blacklight::Document::PageHeaderComponent.new(document: @document, blacklight_config:, search_context: @search_context, search_session:) unless blacklight_config.full_width_header? %>
```
Can we push this class into the config? Can we move the conditional into the component's `render?` method?

dnoneill replied:
No because it is shows up in two different places. If it is full_width we want it to render, just somewhere else.

dnoneill replied:
But yes we can push the class into config.

jcoyne commented on lib/blacklight/configuration.rb:
```diff
@@ -156,6 +156,12 @@ def initialized_default_configuration?
+      # @since v8.3.1
+      # When true, doesn't place the header in col-9 documents but outside as own div
+      # @return [Boolean]
+      property :full_width_header?, default: false
```
I don't think config properties usually have question marks, do they?

jcoyne commented on lib/blacklight/configuration.rb:
```diff
@@ -558,6 +565,22 @@ def add_show_tools_partial(name, opts = {})
+    #   @option opts [String] :partial partial to draw if component is false
+    #   @option opts [Symbol,Proc] :if render this action if the method identified by the symbol or the proc evaluates to true. The proc will receive the action configuration and the document or documents for the action.
+    #   @option opts [Symbol,Proc] :unless render this action unless the method identified by the symbol or the proc evaluates to true. The proc will receive the action configuration and the document or documents for the action.
+    def add_header_tools_partial(name, opts = {})
```
I think we want to make this clear that this is the "show" header.  Not the header on the index view.

dnoneill replied:
how do you like add_show_header_tools_partial?

jcoyne commented on app/assets/stylesheets/blacklight/_bookmark.scss:
```diff
@@ -37,4 +37,8 @@ label.toggle-bookmark {
+}
+
+.header-tools label.toggle-bookmark {
+  min-width: 2rem;
```
What does this style do for us? What happens without it?

dnoneill replied:
The bookmark gets moved really far to the right. I am overriding another css class that I am assuming is to keep tools formatted properly https://github.com/projectblacklight/blacklight/blob/a8b5a3d3a2e8243c8a875d40528eae5c106d83bd/app/assets/stylesheets/blacklight/_bookmark.scss#L3

jcoyne replied:
Can you add a comment into the CSS file so that other people will understand the reasoning for this rule?

jcoyne commented on app/components/blacklight/document/page_header_component.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= render blacklight_config.track_search_session.applied_params_component.new if blacklight_config.track_search_session.applied_params_component %>
```
Can this be moved into a method on the component? Something like `<%= applied_params %>`

dnoneill commented:
@jcoyne I added your suggestions except for the unresolved ones. I also added tests for the PageHeaderComponent and added a show_header_tools_component to the blacklight config.

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,73 @@
+      end
+
+      def before_render
+        @blacklight_config = helpers.blacklight_config
```
Do we need this `before_render` method?  Can we just delegate `blacklight_config` to `helpers`?

dnoneill replied:
We did not. I didn't know you could delegate in components without causing errors. Thanks for the tip.

jcoyne commented on lib/blacklight/configuration.rb:
```diff
@@ -156,6 +156,12 @@ def initialized_default_configuration?
 
+      # @!attribute full_width_header
+      # @since v8.3.1
+      # When true, doesn't place the header in col-9 documents but outside as own div
```
I'm not sure it's clear to say "in col-9 documents".  Can you be more descriptive here?

Is there a reason we need to support this flag? Can we do it just one way?  This sort of configuration makes it challenging to test as it adds a whole new dimension that we should be testing for.  Is it possible for just your application to override the correct components to get the layout as you need?

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,73 @@
+
+        render Blacklight::Document::ActionsComponent.new(document: document,
+                                                          tag: 'div',
+                                                          classes: 'd-inline-flex header-tools align-items-center col-12 col-md-6 ms-auto justify-content-md-end',
```
Extracting these classes (as well as the component class itself) to a separate method makes it easier for people to configure via overriding.

dnoneill replied:
Done.

jcoyne commented on app/assets/stylesheets/blacklight/_bookmark.scss:
```diff
@@ -37,4 +37,10 @@ label.toggle-bookmark {
   }
+}
+
+/* override for line 3.
```
Can we move this rule up by line 3 so that it's easier to see the connection?

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,87 @@
+      end
+
+      def render?
+        search_context.present? || search_session.present? || has_header_tools?
```
Is there a reason we need this method?  Won't `search_session` always exist, which means this is always truthful?  See https://github.com/projectblacklight/blacklight/blob/a8b5a3d3a2e8243c8a875d40528eae5c106d83bd/app/controllers/concerns/blacklight/search_context.rb#L56-L61

dnoneill replied:
Actually present? works for empty hashes, so I changed it back.

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,87 @@
+      def pagination_component
+        return unless blacklight_config.track_search_session.item_pagination_component
+
+        blacklight_config.track_search_session.item_pagination_component.new(search_context: search_context, search_session: search_session, current_document: document)
```
```suggestion
        blacklight_config.track_search_session.item_pagination_component.new(search_context:, search_session:, current_document: document)
```

dnoneill replied:
I did this originally. Rubocop complains.

jcoyne replied:
I see that in some `.erb` files you did this too.  So can we be consistent in the style we're using?

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,87 @@
+      end
+
+      def header_container_classes
+        has_header_tools? ? 'd-flex justify-content-between pagination-search-widgets pb-2 row justify-content-end' : 'pagination-search-widgets'
```
I don't understand why we use both grid classes `row` and flex classes here.  I don't understand why we'd want both `justify-content-between` and `justify-content-end`.  Don't those override each other?

dnoneill replied:
cruft.

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,87 @@
+      end
+
+      def header_actions
+        actions = helpers.send(:filter_partials, blacklight_config.view_config(:show).header_actions, { document: document })
```
Please change the visibility of `filter_partials` or move it in here. Don't use `send` to override the encapsulation.

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,87 @@
+                                                          url_opts: Blacklight::Parameters.sanitize(params.to_unsafe_h))
+      end
+
+      def tag
```
Can you change the name of this method. I suspect it may be confusing with the `tag` heleper method https://api.rubyonrails.org/v5.1.7/classes/ActionView/Helpers/TagHelper.html#method-i-tag

dnoneill replied:
updated to action_component_tag

jcoyne commented on app/components/blacklight/document/page_header_component.rb:
```diff
@@ -0,0 +1,87 @@
+      def render_header_tools
+        return unless has_header_tools?
+
+        show_header_tools_component&.tap do |show_tools_component_class|
```
Is this the same as `return render show_header_tools_component.new(document: document) if show_header_tools_component` ?  I think writing it that way is clearer.

dnoneill replied:
True but this is the way it is in show_tools so I figured we would match? https://github.com/projectblacklight/blacklight/blob/a8b5a3d3a2e8243c8a875d40528eae5c106d83bd/app/components/blacklight/document/sidebar_component.rb#L20-L22
